ros/src/tl_detector/tl_detector.py

```diff
@@ -283,14 +295,31 @@ def process_traffic_lights(self):
                  (specified in styx_msgs/TrafficLight)
 
         """
+
+        # searching_distance_tl parameter sets distance in which
+        # traffic lights will be searched as number of waypoints
+        searching_distance_tl = 120
+
         light = None
-        light_positions = self.config['light_positions']
+        light_wp = -1
         if(self.pose and self.waypoints):
+            # In case that array with traffic light waypoints does not exist,
+            # create it
+            if(self.tl_waypoints == []):
+                self.get_tl_waypoints()
             car_position = self.get_closest_waypoint(self.pose.pose)
-
-        # TODO find the closest visible traffic light (if one exists)
-
-        if light:
+            # Loop thorugh waypoints to find the closest traffic ligt waypoint
+            smallest_tl_distance = 10000
+            for tl_waypoint in self.tl_waypoints:
+                distance_between_waypoints = abs(car_position-tl_waypoint)
+                if(distance_between_waypoints < searching_distance_tl and
+                    distance_between_waypoints < smallest_tl_distance and
+                        car_position < tl_waypoint):
```

Comment:
I just noticed: how is the `<` operator implemented for the class `Pose`?

Reply:
distance_between_waypoints < searching_distance_tl is checking that the traffic light waypoint is in search range (how many waypoints in front of the car will we search)

distance_between_waypoints < smallest_tl_distance checks that this is the closes traffic light.

car_position < tl_waypoint checks that the car is behind the traffic light.

Reply:
Now I realize: `car_position` is actually **not** a _position_ (object of class `Pose`). It's an index to the waypoint (`Pose`) that represents that position. 

I think the naming should reflect this more clearly, both for `car_position` and the function `get_closest_waypoint` (since it's returning an index and not really a waypoint).

Then my question is: what happens when taking a new lap? The `car_position` could be index 10900 and the next traffic light could be index 10, or did I understand wrong?

Reply:
I'll rename it to car_position_wp, this was from Udacity's original source code:
car_position = self.get_closest_waypoint(self.pose.pose)

That concerned me as well. There is no any traffic lights in first 200+ points and I believe that in real life waypoints should always increase. According to the line that I pasted above I believe that this is the implementation that they expect.

Reply:
Then I will rename get_closest_waypoint to get_closest_waypoint_wp as well. Please wait until I post second commit.

Reply:
What does "wp" stand for? Isn't it redundant to write `waypoint_wp`? I would write `waypoint_index` or `waypoint_idx`.

Regarding the previous comment, I think it would still be good to create a function `waypoint_distance` or similar, that gives you the smallest distance between 2 waypoints (considering the wrapping at the end of the waypoints array). Assuming that there are no traffic lights in the first 200+ points is too big of an assumption IMHO.